### DISCUSSION TITLE
2X Snapclients x 1 ESP32

### DISCUSSION THREAD

**Discussion by DerPicknicker** [CarlosDerSeher/snapclient#87] [Ideas]
Hey @CarlosDerSeher ,

this is just a discussion for my idea to create 2 instances of snapclients at one esp32. 

My idea is: 

An amp board has two outputs - each output should be independently used as snapclient because in Multiroom audio almost every speaker is a mono speaker. 

This would reduce the cost and increases the flexibility. Also for the lyrat board. It has stereo output and then you can easily connect 2 speakers and have independent audio outputs...

This is just an idea which I came up with. Because I waste one channel per speaker one board with 2 independent speakers would be great. Then you can install one board and create different areas...

**Comment by CarlosDerSeher** [comment]
Things that come to my mind

Does the software need to mix LR stereo channels to mono or should it be able to chose L or R

> An amp board has two outputs - each output should be independently used as snapclient because in Multiroom audio almost every speaker is a mono speaker.

Why not just make a stereo speaker then? What's the gain of having two mono speakers which, if I understand you right, will be located spacially close to each other and connected by wires to the DAC? Isn't this kind of contradictory to the sole purpose of having a WiFi speaker?

**Reply by DerPicknicker** [reply]
Maybe I was a bit unclear ;-) ... 

The general use case would be that I have one place where all my Amps are placed. And to all speakers I wire one cable. So one esp32 can now power 2 speakers at the same time ;-)

**Reply by CarlosDerSeher** [reply]
Ok I think I still don't get it. Maybe you could do a simple drawing?

Is this an existing system in your house you try to upgrade?

**Reply by DerPicknicker** [reply]
One place where all wires came together:

Like this 

Living Room 

Kitchen 

Bathroom 


All speakers are wired to one place in the storage room. And for each room I would need one amp. But if I can use one esp32 for two rooms I need only the half ;-)

**Reply by CarlosDerSeher** [reply]
For a quick solution you could add some dsp function to mix LR to mono and store this back to the PCM queue. This would result in an i2s stream having the same information in left and right channel achieving exactly what you need without the need of a second snapclient instance. This would be a better approach I think. You could make this configurable through menuconfig.

**Reply by CarlosDerSeher** [reply]
[Here](https://github.com/ashwek/Mixer/blob/master/stereo_to_mono.cpp) is a example of how to do this. A simple average of LR should be good enough to get you started I think. There could be some clipping maybe, not sure how to handle this.

**Reply by DerPicknicker** [reply]
Thank you for the hint. The mono audio can already be achieved with the bridge mode. But this results that only one channel is playing audio (2x power)...

I want that both channels are independent but as you said it's too much work needed.

**Reply by CarlosDerSeher** [reply]
Don't use bridge mode. Just configure your dac as stereo. Mix the audio to mono in software and pack this mono audio to both L and R channel. Voila you can use both channels of your dac and provide 2 rooms with the same audio. Shouldn't be too hard to get this going.

Or did I miss something?

Edit: oh I get it you want to do a different audio stream in another room?

**Reply by DerPicknicker** [reply]
Yes your edit is correct. Two different streams.

**Reply by CarlosDerSeher** [reply]
This will complicate things even more. You'll also have to aggregate the audio data from two snapclients to a single i2s stream.

**Reply by DerPicknicker** [reply]
I think we can stop here ;-) getting more esp32+dac (or a raspberry with multiple dacs)should be much easier. 

Would be a really Cool task for cold and bad days. @CarlosDerSeher are you now motivated 😅 ?

**Reply by CarlosDerSeher** [reply]
Maybe 😀 but this could take a while with my second child on her way. So don't wait for it 😉

**Reply by DerPicknicker** [reply]
No, No... I have firstly get my MA instance running on a new machine... ordering esp32 boards. 

This is why I love the esp32 it doesn't have a file system .. So you can't break it. A power cycle and you're good to go.


Edit: ALL THE BEST FOR YOUR 2nd one 👍🏻

**Comment by CarlosDerSeher** [comment]
Also I can tell you this will be quite an adventure. You'll need to change all those global variables (decoder, audio buffer, etc) to dynamically created ones in the context of a snapclient module. 

This will of course improve the code a lot and clean up most of the mess I created by "quick and dirty" making this thing do synced playback. Do you think you can do this? First things that comes to my mind is a seperate netconn connection not sure if we could just start http_task twice. I don't think so without refactoring big parts of the code.

**Reply by DerPicknicker** [reply]
I thought just double the player tasks and change a bit of the buffers. So I can do it but the problem is if you already say that it will be an Adventure then I think it will be more time consuming for me. It's not a skill issue - just a thing about priorities.. 

Because all I do with embedded is just my personal hobby. So maybe it's easier just get an raspberry and connect multiple soundcards to it. 

If you see any other option to get this ready please let me know.

**Reply by CarlosDerSeher** [reply]
You'll also need a separate TCP connection (http_task) and a separate decoder (maybe others too don't knows right now). Not sure if the processor can handle all this, I think it could handle 2 though. Then there comes a decision, which of the time syncs you'll use. snapclient instance 1 or 2? Then you'll need the player task to aggregate both streams and make sure you pick the same point in time from the buffer. I don't think this will be something you'll come up with just like that. You could maybe just duplicate the player task and put all audio to a ring buffer with the size of one decoded chunk and do the i2s part from another task. Wait until both buffers have enough data and then start building i2s data.

**Reply by CarlosDerSeher** [reply]
> So maybe it's easier just get an raspberry and connect multiple soundcards to it.

Why not get multiple ESP + DAC? Should be cheaper this way or am I wrong?

**Reply by DerPicknicker** [reply]
Lyrat costs around 20.. The esperagus HiFi costs 2x of the Lyrat and I guess it sounds much better due the lower noise.

Lyrat is horrible for audio. I can't find the time to finish the toslink pcb which will result in much lower noise. 

A raspberry pi costs 20$ (for a zero). A usb hub (10$) and a 7ch usb soundcard 15$. 

45$ for 7ch snapclients...
